app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/LocalFileTransferActivity.kt

```diff
@@ -0,0 +1,382 @@
+            arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+            PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS
+          )
+        })
+      } else {
+        ActivityCompat.requestPermissions(
+          this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+          PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS
+        )
+      }
+      false
+    } else {
+      true
+      // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  override fun onRequestPermissionsResult(
+    requestCode: Int,
+    permissions: Array<String>,
+    grantResults: IntArray
+  ) {
+    if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
+      when (requestCode) {
+        PERMISSION_REQUEST_CODE_COARSE_LOCATION -> {
+          Log.e(
+            TAG,
+            "Location permission not granted"
+          )
+          R.string.permission_refused_location.toast(this)
+
+          finish()
+        }
+        PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS -> {
+          Log.e(
+            TAG,
+            "Storage write permission not granted"
+          )
+          R.string.permission_refused_storage.toast(this)
+
+          finish()
+        }
+        else -> {
+          super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        }
+      }
+    }
+  }
+
+  private val isLocationServiceEnabled: Boolean
+    get() = isProviderEnabled(LocationManager.GPS_PROVIDER) ||
+      isProviderEnabled(LocationManager.NETWORK_PROVIDER)
+
+  private fun isProviderEnabled(locationProvider: String): Boolean {
+    return try {
+      locationManager.isProviderEnabled(locationProvider)
+    } catch (ex: SecurityException) {
+      ex.printStackTrace()
+      false
+    } catch (ex: IllegalArgumentException) {
+      ex.printStackTrace()
+      false
+    }
+  }
+
+  private fun requestEnableLocationServices() {
+    alertDialogShower.show(
+      KiwixDialog.EnableLocationServices, {
+        startActivityForResult(
+          Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
+          REQUEST_ENABLE_LOCATION_SERVICES
+        )
+      },
+      R.string.discovery_needs_location.toast(this)::show
+    )
+  }
+
+  private fun requestEnableWifiP2pServices() {
+    alertDialogShower.show(
+      KiwixDialog.EnableWifiP2pServices, {
+        startActivity(Intent(Settings.ACTION_WIFI_SETTINGS))
+      }, R.string.discovery_needs_wifi.toast(this)::show
+    )
+  }
+
+  override fun onActivityResult(
+    requestCode: Int,
+    resultCode: Int,
+    data: Intent?
+  ) {
+    when (requestCode) {
+      REQUEST_ENABLE_LOCATION_SERVICES -> {
+        if (!isLocationServiceEnabled) {
+          R.string.permission_refused_location.toast(this)
+        }
+      }
+      else -> {
+        super.onActivityResult(requestCode, resultCode, data)
+      }
+    }
+  }
+
+  fun Any.toast(context: Context, duration: Int = Toast.LENGTH_SHORT): Toast =
```

Comment:
There is already an extension on Activity called `toast` use that

Also this would produce gibberish, `R.id.whatever` is an int so when you call .toast on them what happens is 
```
123456.toast(...
```

calling toString on an int just creates a string of that number